diff -NwrdBU4 nethack-3.4.3-orig/include/config.h qm/include/config.h
--- nethack-3.4.3-orig/include/config.h	2003-12-07 17:39:14.000000000 -0600
+++ qm/include/config.h	2004-02-11 15:25:42.000000000 -0600
@@ -349,8 +349,9 @@
  */
 
 /*#define GOLDOBJ */	/* Gold is kept on obj chains - Helge Hafting */
 /*#define AUTOPICKUP_EXCEPTIONS */ /* exceptions to autopickup */
+#define WEBB_NAMED_MONSTERS
 
 /* End of Section 5 */
 
 #include "global.h"	/* Define everything else according to choices above */
diff -NwrdBU4 nethack-3.4.3-orig/include/extern.h qm/include/extern.h
--- nethack-3.4.3-orig/include/extern.h	2003-12-07 17:39:14.000000000 -0600
+++ qm/include/extern.h	2004-02-11 15:25:42.000000000 -0600
@@ -320,8 +320,12 @@
 /* ### do_name.c ### */
 
 E int FDECL(getpos, (coord *,BOOLEAN_P,const char *));
 E struct monst *FDECL(christen_monst, (struct monst *,const char *));
+#if defined(WEBB_NAMED_MONSTERS)
+E void FDECL(introduce_mons, (struct monst *));
+E const char * FDECL(monst_rnd_name,(int, int));
+#endif
 E int NDECL(do_mname);
 E struct obj *FDECL(oname, (struct obj *,const char *));
 E int NDECL(ddocall);
 E void FDECL(docall, (struct obj *));
diff -NwrdBU4 nethack-3.4.3-orig/src/bones.c qm/src/bones.c
--- nethack-3.4.3-orig/src/bones.c	2003-12-07 17:39:14.000000000 -0600
+++ qm/src/bones.c	2004-02-11 15:25:42.000000000 -0600
@@ -230,8 +230,12 @@
 	    if (mtmp->iswiz || mptr == &mons[PM_MEDUSA] ||
 		    mptr->msound == MS_NEMESIS || mptr->msound == MS_LEADER ||
 		    mptr == &mons[PM_VLAD_THE_IMPALER])
 		mongone(mtmp);
+#ifdef WEBB_NAMED_MONSTERS
+     else if (mtmp->mnamelth && (uchar)NAME(mtmp)[0]>(uchar)0x80)
+       christen_monst(mtmp, "");
+#endif
 	}
 #ifdef STEED
 	if (u.usteed) dismount_steed(DISMOUNT_BONES);
 #endif
diff -NwrdBU4 nethack-3.4.3-orig/src/do_name.c qm/src/do_name.c
--- nethack-3.4.3-orig/src/do_name.c	2003-12-07 17:39:14.000000000 -0600
+++ qm/src/do_name.c	2004-02-11 15:25:42.000000000 -0600
@@ -220,8 +220,19 @@
 	replmon(mtmp,mtmp2);
 	return(mtmp2);
 }
 
+#if defined(WEBB_NAMED_MONSTERS)
+void
+introduce_mons(mtmp)
+struct monst * mtmp;
+{
+  char * name = NAME(mtmp);
+  if (name && (uchar) name[0] >= (uchar)0x80)
+     name[0] &= 0x7F;
+}
+#endif
+
 int
 do_mname()
 {
 	char buf[BUFSZ];
@@ -547,8 +558,72 @@
 {
     return rn2(7) ? ghostnames[rn2(SIZE(ghostnames))] : (const char *)plname;
 }
 
+#if defined(WEBB_NAMED_MONSTERS)
+static char * erinysnames[] = {
+  "\301lecto",
+  "\315egaera",
+  "\324isiphone"
+};
+
+static int erinys_shuffled = FALSE;
+  
+static const char * qmnames[] = {
+  "\302uckaroo",  /* B. Banzai */
+  "\304ick",      /* D. Feynman */
+  "\305rwin",     /* E. Schroedinger */
+  "\314ouis",     /* L. de Broglie */
+  "\315ax",       /* M. Born, M. Planck */
+  "\316iels",     /* N. Bohr */
+  "\320ascual",   /* P. Jordan */
+  "\320aul",      /* P. Dirac */
+  "\323am",       /* S. Beckett */ 
+  "\323atyen",    /* Satyendra Bose */
+  "\323ephen",    /* S. Hawking */
+  "\327erner",    /* W. Heisenberg */
+  "\327olfgang",  /* W. Pauli */
+  "\305nrico",    /* E. Fermi */  
+
+  "\315arie",     /* M. Curie */
+  "\314ise",      /* L. Meitner */
+};
+
+
+const char *
+monst_rnd_name(mtype, gender)
+int mtype, gender;
+{
+  char * empty  = "";
+  switch (mtype){
+    case PM_QUANTUM_MECHANIC:
+      if (gender) 
+        return (rn2(3))?empty:qmnames[SIZE(qmnames)-rnd(2)];
+      return ((rn2(3))?qmnames[rn2(SIZE(qmnames)-2)]:empty);
+      break;
+    case PM_ERINYS:
+      if (mvitals[PM_ERINYS].born > 3) return empty;
+      else if (!erinys_shuffled){
+        char * swap;
+        int index, i, seed = (u.ubirthday/10)%6; /* 6 = 3! */
+        for (i = 2; i > 0; --i){
+          index = seed%(i+1);
+          seed /= (i+1); 
+          swap = erinysnames[i];
+          erinysnames[i] = erinysnames[index];
+          erinysnames[index] = swap;
+        }
+        erinys_shuffled = TRUE;
+      }
+      return erinysnames[ mvitals[PM_ERINYS].born - 1 ];
+      break;
+    default:
+      return empty;
+  }
+}
+
+#endif
+
 /* Monster naming functions:
  * x_monnam is the generic monster-naming function.
  *		  seen	      unseen	   detected		  named
  * mon_nam:	the newt	it	the invisible orc	Fido
@@ -678,11 +751,18 @@
 
 	/* Put the actual monster name or type into the buffer now */
 	/* Be sure to remember whether the buffer starts with a name */
 	if (do_hallu) {
+
+
+
 	    Strcat(buf, rndmonnam());
 	    name_at_start = FALSE;
+#if defined(WEBB_NAMED_MONSTERS)
+  } else if (mtmp->mnamelth && (uchar)NAME(mtmp)[0]<(uchar)0x80) {     
+#else
 	} else if (mtmp->mnamelth) {
+#endif
 	    char *name = NAME(mtmp);
 
 	    if (mdat == &mons[PM_GHOST]) {
 		Sprintf(eos(buf), "%s ghost", s_suffix(name));
diff -NwrdBU4 nethack-3.4.3-orig/src/end.c qm/src/end.c
--- nethack-3.4.3-orig/src/end.c	2003-12-07 17:39:14.000000000 -0600
+++ qm/src/end.c	2004-02-11 15:25:42.000000000 -0600
@@ -219,11 +219,16 @@
 		killer = m_monnam(mtmp);
 		Strcat(buf, killer);
 	} else {
 		Strcat(buf, mtmp->data->mname);
-		if (mtmp->mnamelth)
+      if (mtmp->mnamelth
+#if defined(WEBB_NAMED_MONSTERS)
+          && (uchar)NAME(mtmp)[0]<(uchar)0x80
+#endif
+          ) {     
 		    Sprintf(eos(buf), " called %s", NAME(mtmp));
 	}
+	}
 
 	if (multi) Strcat(buf, ", while helpless");
 	killer = buf;
 	if (mtmp->data->mlet == S_WRAITH)
diff -NwrdBU4 nethack-3.4.3-orig/src/makemon.c qm/src/makemon.c
--- nethack-3.4.3-orig/src/makemon.c	2003-12-07 17:39:14.000000000 -0600
+++ qm/src/makemon.c	2004-02-11 15:25:42.000000000 -0600
@@ -620,14 +620,29 @@
 	    case S_MUMMY:
 		if (rn2(7)) (void)mongets(mtmp, MUMMY_WRAPPING);
 		break;
 	    case S_QUANTMECH:
+#if defined(WEBB_NAMED_MONSTERS)
+      {
+        char * ame = (mtmp->mnamelth &&
+            (uchar)NAME(mtmp)[0] > (uchar)0x80)?NAME(mtmp)+1:0;
+        if (ptr == &mons[PM_QUANTUM_MECHANIC]){
+          if (!rn2(20) || (ame && !strcmp(ame,"rwin") && rn2(20))) {
+            otmp = mksobj(LARGE_BOX, FALSE, FALSE);
+            otmp->spe = 1; /* flag for special box */
+            otmp->owt = weight(otmp);
+            (void) mpickobj(mtmp, otmp);
+          }
+        }
+      }
+#else
 		if (!rn2(20)) {
 			otmp = mksobj(LARGE_BOX, FALSE, FALSE);
 			otmp->spe = 1; /* flag for special box */
 			otmp->owt = weight(otmp);
 			(void) mpickobj(mtmp, otmp);
 		}
+#endif
 		break;
 	    case S_LEPRECHAUN:
 #ifndef GOLDOBJ
 		mtmp->mgold = (long) d(level_difficulty(), 30);
@@ -1020,8 +1035,15 @@
 	} else if (mndx == PM_GHOST) {
 		flags.ghost_count++;
 		if (!(mmflags & MM_NONAME))
 			mtmp = christen_monst(mtmp, rndghostname());
+#ifdef WEBB_NAMED_MONSTERS
+   } else if (mndx == PM_QUANTUM_MECHANIC ) {
+     if(!(mmflags & MM_NONAME))
+       mtmp = christen_monst(mtmp, monst_rnd_name(mndx, mtmp->female));
+   } else if (mndx == PM_ERINYS && countbirth){ /* ignore MM_NONAME */
+     mtmp = christen_monst(mtmp, monst_rnd_name(mndx, 1));
+#endif
 	} else if (mndx == PM_VLAD_THE_IMPALER) {
 		mitem = CANDELABRUM_OF_INVOCATION;
 	} else if (mndx == PM_CROESUS) {
 		mitem = TWO_HANDED_SWORD;
diff -NwrdBU4 nethack-3.4.3-orig/src/mon.c qm/src/mon.c
--- nethack-3.4.3-orig/src/mon.c	2003-12-07 17:39:14.000000000 -0600
+++ qm/src/mon.c	2004-02-11 15:25:42.000000000 -0600
@@ -298,9 +298,13 @@
 	/* if polymorph or undead turning has killed this monster,
 	   prevent the same attack beam from hitting its corpse */
 	if (flags.bypasses) bypass_obj(obj);
 
-	if (mtmp->mnamelth)
+  if (mtmp->mnamelth
+#if defined(WEBB_BIODIVERSITY) || defined(WEBB_NAMED_MONSTERS)
+      && (uchar)NAME(mtmp)[0] < (uchar) 0x80
+#endif
+     )
 	    obj = oname(obj, NAME(mtmp));
 
 	/* Avoid "It was hidden under a green mold corpse!" 
 	 *  during Blind combat. An unseen monster referred to as "it"
@@ -1640,8 +1644,13 @@
 		   so that saved monster traits won't retain any stale
 		   item-conferred attributes */
 		otmp = mkcorpstat(STATUE, KEEPTRAITS(mdef) ? mdef : 0,
 				  mdef->data, x, y, FALSE);
+		if (mdef->mnamelth
+#if defined(WEBB_BIODIVERSITY) || defined(WEBB_NAMED_MONSTERS)
+                     && ((uchar)NAME(mdef)[0]) < (uchar)0x80
+#endif
+          ) otmp = oname(otmp, NAME(mdef));
 		if (mdef->mnamelth) otmp = oname(otmp, NAME(mdef));
 		while ((obj = oldminvent) != 0) {
 		    oldminvent = obj->nobj;
 		    (void) add_to_container(otmp, obj);
diff -NwrdBU4 nethack-3.4.3-orig/src/pager.c qm/src/pager.c
--- nethack-3.4.3-orig/src/pager.c	2003-12-07 17:39:14.000000000 -0600
+++ qm/src/pager.c	2004-02-11 15:25:42.000000000 -0600
@@ -126,8 +126,14 @@
 	mtmp = m_at(x,y);
 	if (mtmp != (struct monst *) 0) {
 	    char *name, monnambuf[BUFSZ];
 	    boolean accurate = !Hallucination;
+#if defined(WEBB_NAMED_MONSTERS)
+      /* You read their mind, you learn an unknown name */
+      if (mtmp->mnamelth  && (uchar)(name = NAME(mtmp))[0] >= (uchar)0x80 && 
+          accurate && (tp_sensemon(mtmp)) )
+        introduce_mons(mtmp);
+#endif
 
 	    if (mtmp->data == &mons[PM_COYOTE] && accurate)
 		name = coyotename(mtmp, monnambuf);
 	    else
diff -NwrdBU4 nethack-3.4.3-orig/src/restore.c qm/src/restore.c
--- nethack-3.4.3-orig/src/restore.c	2003-12-07 17:39:14.000000000 -0600
+++ qm/src/restore.c	2004-02-11 15:25:42.000000000 -0600
@@ -276,8 +276,12 @@
 			if (propagate(mndx, TRUE, ghostly) == 0) {
 				/* cookie to trigger purge in getbones() */
 				mtmp->mhpmax = DEFUNCT_MONSTER;	
 			}
+#if defined(WEBB_NAMED_MONSTERS)
+         else if (mndx == PM_ERINYS) /* ignore MM_NONAME */
+           mtmp = christen_monst(mtmp, monst_rnd_name(mndx, 1));
+#endif
 		}
 		if(mtmp->minvent) {
 			struct obj *obj;
 			mtmp->minvent = restobjchn(fd, ghostly, FALSE);
diff -NwrdBU4 nethack-3.4.3-orig/src/sounds.c qm/src/sounds.c
--- nethack-3.4.3-orig/src/sounds.c	2003-12-07 17:39:14.000000000 -0600
+++ qm/src/sounds.c	2004-02-11 15:25:42.000000000 -0600
@@ -669,8 +669,22 @@
 	    if (!mtmp->mpeaceful) {
 		if (In_endgame(&u.uz) && is_mplayer(ptr)) {
 		    mplayer_talk(mtmp);
 		    break;
+#if defined(WEBB_NAMED_MONSTERS)
+        } else if (mtmp->mnamelth && (uchar)NAME(mtmp)[0] >= (uchar)0x80){
+          introduce_mons(mtmp);
+          if (ptr == &mons[PM_QUANTUM_MECHANIC])
+            verbalize("I'm %s, are you ready for the quantum immortality experiment?", NAME(mtmp));
+          else
+            verbalize("Hello, my name is %s. Prepare to die.", NAME(mtmp));
+            /* Princess Bride*/
+          break;
+        } else if (ptr == &mons[PM_QUANTUM_MECHANIC] && 
+            !strcmp(NAME(mtmp), "Sam")){
+          verbalize("Oh, boy.");
+          break;
+#endif
 		} else return 0;	/* no sound */
 	    }
 	    /* Generic peaceful humanoid behaviour. */
 	    if (mtmp->mflee)
@@ -691,8 +705,20 @@
 	    else if (mtmp->mtame && !mtmp->isminion &&
 						moves > EDOG(mtmp)->hungrytime)
 		verbl_msg = "I'm hungry.";
 	    /* Specific monsters' interests */
+#if defined(WEBB_NAMED_MONSTERS)
+      else if (mtmp->mnamelth && (uchar)NAME(mtmp)[0] >= (uchar)0x80){
+        introduce_mons(mtmp);
+        verbalize("%s, I'm %s.  %s", Hello(mtmp), NAME(mtmp), 
+            (ptr == &mons[PM_QUANTUM_MECHANIC])?
+            "Please come back during my office hours.":
+            (Inhell)?"Hot enough for ya?":"Dank enough for ya?");
+      } else if (ptr == &mons[PM_QUANTUM_MECHANIC] && 
+          !strcmp(NAME(mtmp), "Sam")){
+        verbl_msg = "Oh, boy.";
+      }
+#endif
 	    else if (is_elf(ptr))
 		pline_msg = "curses orcs.";
 	    else if (is_dwarf(ptr))
 		pline_msg = "talks about mining.";
